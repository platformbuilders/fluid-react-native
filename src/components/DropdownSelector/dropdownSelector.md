
# Dropdown Selector âŒ¨ 

## Example

| android | -iOS- |
| :-------------: | :-------------: |
| ![image](https://user-images.githubusercontent.com/84458435/129288024-fad35708-4ebd-49dc-a78f-6194e781c9c6.png) | ![image](https://user-images.githubusercontent.com/84458435/129301120-07726b5f-6c91-4135-9e90-828e78a06080.png)
 |




## Source Code With Forkik
```js
import React from 'react';
import { DropdownSelector } from '@platformbuilders/react-native-ui';

const Home: React.FC = () => {
  return (
          <DropdownSelector
            options={[
              { label: 'a', value: 'b' },
              { label: 'Example Dropdown' },
            ]}
            selectedValue={values?.myField}
            setSelectedValue={handleChange('myField')}
            id="lg_my_field_input"
            accessibility="Example"
            label="Example"
            iconName="rocket"
            leftIconName="rocket"
            error='error example'
          />
  );
}

export default Home;
```
## Source Code With States
```js
import React from 'react';
import { DropdownSelector } from '@platformbuilders/react-native-ui';

const Home: React.FC = () => {
  return (
                <DropdownSelector
            options={[
              { label: 'a', value: 'b' },
              { label: 'Example Dropdown' },
            ]}
            selectedValue={myState}
            setSelectedValue={setMyState}
            id="lg_my_field_input"
            accessibility="Example"
            label="Example"
            iconName="rocket"
            leftIconName="rocket"
            error='error example'
          />
  );
}

export default Home;
```





