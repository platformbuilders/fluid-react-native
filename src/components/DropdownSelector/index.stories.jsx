import { storiesOf } from '@storybook/react-native';
import dropdownSelectorMarkdown from './dropdownSelector.md';

storiesOf('Text Input', module)
  .addParameters({
    readme: {
      content: dropdownSelectorMarkdown,
    },
    viewport: {
      defaultViewport: '',
    },
  })
  .add('Example', () => {});
