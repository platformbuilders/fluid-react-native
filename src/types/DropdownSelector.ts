import { PickerItemProps, PickerProps } from '@react-native-picker/picker';
import { HitSlopType } from './Common';
import { FontType } from './IconType';
import { TypographyVariants } from './Variants';

export type DropdownSelectorType = {
  // component content
  id: string;
  accessibility: string;
  options: PickerItemProps[];
  // layout
  large?: boolean;
  contrast?: boolean;
  centered?: boolean;
  iconSize?: number;
  iconTouchableEnabled?: boolean;
  status?: string;
  error?: string | boolean;
  iconHitSlop?: HitSlopType;
  iconName?: string;
  label?: string;
  style?: any;
  leftIconColor?: string;
  iconColor?: string;
  borderedBackgroundColor?: string;
  borderedHeight?: number;
  borderedColor?: string;
  borderedRadius?: number;
  fixedLabelVariant?: TypographyVariants;
  rightIconName?: string;
  leftIconName?: string;
  iconType?: FontType;
  // state handlers
  onPressIcon?(arg?: any): void;
  onRightIconPress?(x?: any): void;
  selectedValue?: any;
  setSelectedValue?(itemValue?: any, itemIndex?: number): void;
} & PickerProps;
