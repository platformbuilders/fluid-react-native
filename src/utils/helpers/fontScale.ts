import { RFValue } from 'react-native-responsive-fontsize';

export const scaledFontSize = (
  size: number | string,
  baselineFactor?: 280,
): number => {
  const value = typeof size === 'string' ? parseInt(size, 10) : size;
  return RFValue(value, baselineFactor);
};
