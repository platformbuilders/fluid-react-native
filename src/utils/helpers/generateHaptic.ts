import Haptic, { HapticFeedbackType } from 'react-native-haptic';
import { isIOS } from './isIOS';

export const generateHaptic = (haptic: HapticFeedbackType = 'impact'): void => {
  if (isIOS()) {
    Haptic.generate(haptic);
  }
};
