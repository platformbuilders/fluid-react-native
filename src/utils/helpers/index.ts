export {
  currencyParser,
  parseToThousands,
  removeWhiteSpaces,
  toOnlyNumbers,
  getTheme,
  ifStyle,
  switchStyle,
} from '@platformbuilders/helpers';
export * from './generateHaptic';
export * from './getByTypographyVariant';
export * from './fontScale';
export * from './shadow';
