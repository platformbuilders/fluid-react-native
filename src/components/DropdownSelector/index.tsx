import React, { FC } from 'react';
import { isEmpty } from 'lodash';
import { moderateScale } from 'react-native-size-matters';
import { Picker } from '@react-native-picker/picker';
import { DropdownSelectorType } from '../../types';

import FormError from '../FormError';
import {
  BorderedWrapper,
  FixedLabel,
  Icon,
  InputBorderedAreaWrapper,
  InputBorderedColumnWrapper,
  Wrapper,
} from './styles';

const DropdownSelector: FC<DropdownSelectorType> = ({
  id,
  accessibility,
  large = false,
  contrast = false,
  centered = false,
  iconSize = 20,
  iconTouchableEnabled = false,
  iconName,
  label = '',
  error = '',
  style = {},
  iconHitSlop = {},
  onPressIcon = (): any => {},
  onRightIconPress = (): any => {},
  rightIconName,
  leftIconName = '',
  leftIconColor,
  iconColor,
  borderedBackgroundColor,
  borderedHeight = moderateScale(50),
  borderedColor,
  borderedRadius,
  iconType = 'material',
  fixedLabelVariant = 'caption2',
  selectedValue,
  setSelectedValue = (): void => {},
  options,
  testID,
  ...rest
}) => {
  const hasError = !isEmpty(error);

  const _renderPickerInput = (): JSX.Element => {
    return (
      <Picker
        testID={testID || id || accessibility}
        selectedValue={selectedValue}
        onValueChange={(itemValue: any, itemIndex: number) => {
          setSelectedValue(itemValue, itemIndex);
        }}
        mode="dropdown"
        {...rest}
      >
        {options.map((option) => (
          <Picker.Item
            label={option.label}
            key={option.value || option.label}
            value={option.value || option.label}
            {...option}
          />
        ))}
      </Picker>
    );
  };

  const _renderIcon = (iconProp: string, isLeft?: boolean) => (
    <Icon
      type={iconType}
      id={`id_${iconProp}`}
      accessibility={`icon_${accessibility}`}
      size={iconSize}
      name={iconProp || ''}
      contrast={contrast}
      error={hasError}
      touchable={iconTouchableEnabled}
      onPress={!!rightIconName ? onRightIconPress : onPressIcon}
      hitSlop={iconHitSlop}
      iconColor={isLeft ? leftIconColor : iconColor}
    />
  );

  return (
    <Wrapper style={style}>
      <FormError
        id={id || accessibility}
        accessibility={accessibility}
        centered={centered}
        error={error}
        large={large}
      >
        <BorderedWrapper
          borderedBackgroundColor={borderedBackgroundColor}
          borderedHeight={borderedHeight}
          borderedColor={borderedColor}
          borderedRadius={borderedRadius}
          error={hasError}
        >
          <InputBorderedAreaWrapper>
            {!isEmpty(leftIconName) && _renderIcon(leftIconName, true)}
            <InputBorderedColumnWrapper hasLeftIcon={!isEmpty(leftIconName)}>
              <FixedLabel
                hasLeftIcon={!isEmpty(leftIconName)}
                variant={fixedLabelVariant}
              >
                {label}
              </FixedLabel>
              {_renderPickerInput()}
            </InputBorderedColumnWrapper>
            {!isEmpty(iconName) && _renderIcon(iconName || '')}
          </InputBorderedAreaWrapper>
        </BorderedWrapper>
      </FormError>
    </Wrapper>
  );
};

export default DropdownSelector;
