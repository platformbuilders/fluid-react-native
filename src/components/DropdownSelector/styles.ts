import styled from 'styled-components/native';
import { getTheme, ifStyle } from '../../utils/helpers';
import DefaultIcon from '../Icon';
import Typography from '../Typography';

type BorderedWrapperProps = {
  borderedBackgroundColor?: string;
  borderedHeight?: number;
  borderedColor?: string;
  borderedRadius?: number;
  error?: boolean;
  borderWidth?: number;
};

type InputBorderedColumnWrapperProps = {
  hasLeftIcon?: boolean;
  multiline?: boolean;
  padding?: number;
};

type FixedLabelProps = {
  hasLeftIcon?: boolean;
};

const hasLabel = ifStyle('label');
const hasError = ifStyle('error');
const isContrast = ifStyle('contrast');
const primaryContrast = getTheme('primary.contrast');
const primaryMain = getTheme('primary.main');
const minimumSpacing = getTheme('minimumSpacing');
const smallSpacing = getTheme('smallSpacing');
const mediumSpacing = getTheme('mediumSpacing');
const failure = getTheme('failure');

type WrapperProps = {
  multiline?: boolean;
  style?: any;
};

export const Wrapper = styled.View<WrapperProps>`
  justify-content: ${hasLabel('flex-end', 'flex-start')};
  padding-top: ${hasLabel(smallSpacing, 0)};
  position: relative;
`;

export const BorderedWrapper = styled.View<BorderedWrapperProps>`
  ${({
    borderedBackgroundColor,
    borderedColor,
    borderedHeight,
    borderedRadius,
    error,
    borderWidth,
    ...rest
  }: BorderedWrapperProps) => {
    const borderedStyle = `
      justify-content: center;
      border: ${borderedColor ? borderWidth || 1 : 0}px solid ${
      error ? failure(rest) : borderedColor || primaryMain(rest)
    };
      background-color: ${borderedBackgroundColor || 'transparent'};
      height: ${borderedHeight}px;
      border-radius: ${borderedRadius ? borderedRadius : 0}px;
    `;

    return `
    border: 0;
    ${borderedStyle}
  `;
  }}
`;

export const BorderedAreaWrapper = styled.View`
  flex-direction: row;
  align-items: center;
  width: 100%;
`;

export const BorderedColumnWrapper = styled.View<InputBorderedColumnWrapperProps>`
  flex-direction: column;
  flex: 1;
`;

export const FixedLabel = styled(Typography)<FixedLabelProps>`
  color: ${primaryMain};
  margin-bottom: ${minimumSpacing};
  margin-top: ${minimumSpacing};
  margin-left: ${mediumSpacing};
  position: absolute;
`;

type IconProps = {
  contrast: boolean;
  error: boolean;
  iconColor?: string;
};

export const Icon = styled(DefaultIcon).attrs((props: IconProps) => ({
  color: hasError(
    failure(props),
    props.iconColor ||
      isContrast(primaryContrast(props), primaryMain(props))(props),
  )(props),
}))<IconProps>``;
