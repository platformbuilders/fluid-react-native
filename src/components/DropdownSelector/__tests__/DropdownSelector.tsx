import 'jest';
import React, { FC, useState } from 'react';
// import React from 'react';
import { fireEvent, render } from 'react-native-testing-library';
import renderer from 'react-test-renderer';
import { ThemeProvider } from 'styled-components/native';
import DropdownSelector from '..';
import { theme } from '../../../test/helpers';

describe('<DropdownSelector />', () => {
  it('should render DropdownSelector', () => {
    const wrapper = renderer.create(
      <ThemeProvider theme={theme}>
        <DropdownSelector
          id="test"
          accessibility=""
          options={[{ label: 'a', value: 'b' }, { label: 'c' }]}
        />
      </ThemeProvider>,
    );

    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it('should render dropdown and trigger state change', () => {
    const options = [
      { label: 'labelA', value: 'valueA' },
      { label: 'labelWithoutValue' },
    ];

    const setStateValue = jest.fn();

    const Component: FC = (): JSX.Element => {
      const [value] = useState('');

      return (
        <ThemeProvider theme={theme}>
          <DropdownSelector
            id="testing_dropdownSelector"
            accessibility=""
            options={options}
            iconName="access-point"
            selectedValue={value}
            setSelectedValue={setStateValue}
          />
        </ThemeProvider>
      );
    };

    const { getByTestId } = render(<Component />);

    const component = getByTestId('testing_dropdownSelector');
    fireEvent(component, 'onValueChange', options[1]);
    expect(setStateValue).toHaveBeenCalledWith(
      { label: options[1].label },
      undefined,
    );
  });
});
