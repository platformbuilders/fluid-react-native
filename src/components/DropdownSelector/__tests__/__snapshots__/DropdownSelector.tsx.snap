// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DropdownSelector /> should render DropdownSelector 1`] = `
<View
  style={
    Array [
      Object {
        "justifyContent": "flex-start",
        "paddingTop": 0,
        "position": "relative",
      },
      Object {},
    ]
  }
>
  <View
    borderedHeight={78.57142857142857}
    error={false}
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
          "borderColor": "#C6B09E",
          "borderRadius": 0,
          "borderStyle": "solid",
          "borderWidth": 0,
          "height": 78.57142857142857,
          "justifyContent": "center",
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "width": "100%",
          },
        ]
      }
    >
      <View
        hasLeftIcon={false}
        style={
          Array [
            Object {
              "flexBasis": 0,
              "flexDirection": "column",
              "flexGrow": 1,
              "flexShrink": 1,
            },
          ]
        }
      >
        <Text
          accessibility=""
          accessibilityLabel=""
          hasLeftIcon={false}
          style={
            Array [
              Object {
                "color": "#212121",
                "fontSize": 11,
                "lineHeight": 13,
              },
              Object {
                "color": "#C6B09E",
                "marginBottom": 6.285714285714286,
                "marginLeft": 31.428571428571427,
                "marginTop": 6.285714285714286,
                "position": "absolute",
              },
            ]
          }
          testID=""
          variant="caption2"
        >
          
        </Text>
        <View>
          <RNCPicker
            items={
              Array [
                Object {
                  "label": "a",
                  "testID": undefined,
                  "textColor": undefined,
                  "value": "b",
                },
                Object {
                  "label": "c",
                  "testID": undefined,
                  "textColor": undefined,
                  "value": "c",
                },
              ]
            }
            onChange={[Function]}
            selectedIndex={0}
            style={
              Array [
                Object {
                  "height": 216,
                },
                undefined,
              ]
            }
            testID="test"
          />
        </View>
      </View>
    </View>
  </View>
</View>
`;
